lib/network_manager.go

```diff
@@ -0,0 +1,787 @@
+
+	// Find the RemoteNode associated with the connection. It should almost always exist, since we create the RemoteNode
+	// as we're attempting to connect to the address.
+	id := NewRemoteNodeId(oc.attemptId)
+	rn := nm.rnManager.GetRemoteNodeById(id)
+	if rn != nil {
+		nm.rnManager.Disconnect(rn)
+	}
+	oc.Close()
+	nm.cmgr.RemoveAttemptedOutboundAddrs(oc.address)
+}
+
+// ###########################
+// ## Persistent Connections
+// ###########################
+
+// refreshConnectIps is called periodically by the persistent connector. It is responsible for connecting to all
+// persistent IP addresses that we are not already connected to.
+func (nm *NetworkManager) refreshConnectIps() {
+	// Connect to addresses passed via the --connect-ips flag. These addresses are persistent in the sense that if we
+	// disconnect from one, we will try to reconnect to the same one.
+	for _, connectIp := range nm.connectIps {
+		if _, ok := nm.persistentIpToRemoteNodeIdsMap.Get(connectIp); ok {
+			continue
+		}
+
+		glog.Infof("NetworkManager.initiatePersistentConnections: Connecting to connectIp: %v", connectIp)
+		id, err := nm.CreateNonValidatorPersistentOutboundConnection(connectIp)
+		if err != nil {
+			glog.Errorf("NetworkManager.initiatePersistentConnections: Problem connecting "+
+				"to connectIp %v: %v", connectIp, err)
+			continue
+		}
+
+		nm.persistentIpToRemoteNodeIdsMap.Set(connectIp, id)
+	}
+}
+
+// ###########################
+// ## Validator Connections
+// ###########################
+
+// SetActiveValidatorsMap is called by the owner of the NetworkManager to update the activeValidatorsMap. This should
+// generally be done whenever the active validators set changes.
+func (nm *NetworkManager) SetActiveValidatorsMap(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator]) {
+	nm.activeValidatorsMapLock.Lock()
+	defer nm.activeValidatorsMapLock.Unlock()
+	nm.activeValidatorsMap = activeValidatorsMap.Clone()
+
+}
+
+func (nm *NetworkManager) getActiveValidatorsMap() *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator] {
+	nm.activeValidatorsMapLock.RLock()
+	defer nm.activeValidatorsMapLock.RUnlock()
+	return nm.activeValidatorsMap.Clone()
+}
+
+// refreshValidatorIndex re-indexes validators based on the activeValidatorsMap. It is called periodically by the
+// validator connector.
+func (nm *NetworkManager) refreshValidatorIndex(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator]) {
+	// De-index inactive validators. We skip any checks regarding RemoteNodes connection status, nor do we verify whether
+	// de-indexing the validator would result in an excess number of outbound/inbound connections. Any excess connections
+	// will be cleaned up by the NonValidator connector.
+	validatorRemoteNodeMap := nm.rnManager.GetValidatorIndex().ToMap()
```

Comment:
Feels like there's a race conditoin here. If activeValidatorsMap is modified, eg during an epoch transitoin, at the same time as refreshValidatorIndex is called, then you could wind up with a weird set of validators in the remoteNodeManager indexes.

Reply:
Nevermind I see that you clone the map with an RLock before this function. Confusing that it's a ConcurrentMap since you don't need it to be here. Leave a comment explaining the race condition.